hypertrace-view-generator/hypertrace-view-generator/src/main/java/org/hypertrace/viewgenerator/generators/SpanEventViewGenerator.java

```diff
@@ -140,21 +143,93 @@ private Event getApiEntrySpanForExitSpan(
    * api_node to api_entry_span (child of api_exit_span) in another api_node 2. exit calls to
    * backend from api_exit_span in api_node
    */
-  Map<ByteBuffer, Integer> computeApiExitCallCount(StructuredTrace trace) {
+  Map<ByteBuffer, ApiExitCallInfo> computeApiExitCallCount(StructuredTrace trace) {
     ApiTraceGraph apiTraceGraph = ViewGeneratorState.getApiTraceGraph(trace);
     // event -> api exit call count for the corresponding api_node
-    Map<ByteBuffer, Integer> eventToApiExitCallCount = Maps.newHashMap();
+    Map<ByteBuffer, ApiExitCallInfo> eventToExitInfo = Maps.newHashMap();
 
     for (ApiNode<Event> apiNode : apiTraceGraph.getApiNodeList()) {
       List<Event> backendExitEvents =
           apiTraceGraph.getExitBoundaryEventsWithNoOutboundEdgeForApiNode(apiNode);
       List<ApiNodeEventEdge> edges = apiTraceGraph.getOutboundEdgesForApiNode(apiNode);
       int totalExitCallCount = backendExitEvents.size() + edges.size();
-      apiNode
-          .getEvents()
-          .forEach(e -> eventToApiExitCallCount.put(e.getEventId(), totalExitCallCount));
+      ApiExitCallInfo apiExitCallInfo = new ApiExitCallInfo().withExitCallCount(totalExitCallCount);
+      edges.forEach(edge -> apiExitCallInfo.handleApiNodeEdge(trace, edge));
+      backendExitEvents.forEach(apiExitCallInfo::handleBackend);
+      apiNode.getEvents().forEach(e -> eventToExitInfo.put(e.getEventId(), apiExitCallInfo));
+    }
+    return eventToExitInfo;
+  }
+
+  static class ApiExitCallInfo {
+
+    private static final String UNKNOWN_SERVICE = "unknown-service";
+    private static final String UNKNOWN_BACKEND = "unknown-backend";
+
+    private final Map<String, String> calleeIdToName;
+    private final Map<String, AtomicInteger> calleeIdToCallCount;
+
+    private int exitCallCount;
+    private int unknownServiceExits;
+    private int unknownBackendExits;
+
+    public ApiExitCallInfo() {
+      unknownServiceExits = 0;
+      unknownBackendExits = 0;
+      this.calleeIdToName = Maps.newHashMap();
+      this.calleeIdToCallCount = Maps.newHashMap();
+    }
+
+    public ApiExitCallInfo withExitCallCount(int exitCallCount) {
+      this.exitCallCount = exitCallCount;
+      return this;
+    }
+
+    void handleApiNodeEdge(StructuredTrace trace, ApiNodeEventEdge edge) {
+      Event event = trace.getEventList().get(edge.getTgtEventIndex());
+      if (EnrichedSpanUtils.containsServiceId(event)) {
```

Comment:
Why not just have `calleeNameToCallCount` map?

Reply:
The current structure allows for exposing the corresponding entity id as well if there is a need (for linking entities in ui in future).
Now, that I had kept ids, using them as map key vs name seemed clean.

Reply:
Could it happen that the name is null, but the id is still present? If yes, we should handle that then while adding the name in the map

Reply:
Yes, that was leap of faith which I shouldn't have taken 😅 . I will handle it

Reply:
If you check If you see here, https://github.com/hypertrace/hypertrace-ingester/blob/main/hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/DefaultServiceEntityEnricher.java  
we are populating serviceId only when serviceName is present so serviceName won't be null when serviceId is present.